### PR TITLE
add a 501 helper to allow override of default response

### PR DESCRIPTION
**Pull request by leejo**
specifically this will allow you to return whatever you like in the
case of an endpoint not being implemented, for example mocking of
response data. add a test case to cover this addition

### CODE REVIEW

File: lib/Mojolicious/Plugin/OpenAPI.pm [hunk 1]
```diff
@@ -25,6 +25,7 @@ sub register {
     $app->helper('openapi.valid_input' => sub { _validate($_[0]) ? undef : $_[0] });
     $app->helper('openapi.spec'        => \&_helper_spec);
     $app->helper('reply.openapi'       => \&_reply);
+    $app->helper('openapi.not_implemented' => $config->{not_implemented} || \&NOT_IMPLEMENTED);
```
Comment (jhthorsen) [inline lib/Mojolicious/Plugin/OpenAPI.pm]: There's no reason for this. The user can just override the helper from the app after loading the plugin.

File: lib/Mojolicious/Plugin/OpenAPI.pm [hunk 2]
```diff
@@ -110,7 +111,8 @@ sub _before_render {
   return unless $has_spec or grep { $path =~ /^$_/ } @{$c->stash('openapi.base_paths')};
 
   my $format = $c->stash('format') || 'json';
-  my $res = $args->{exception} ? EXCEPTION() : !$has_spec ? NOT_FOUND() : NOT_IMPLEMENTED();
+  my $res = $args->{exception} ? EXCEPTION() : !$has_spec
+    ? NOT_FOUND() : $c->openapi->not_implemented($has_spec);
```
Comment (jhthorsen) [inline lib/Mojolicious/Plugin/OpenAPI.pm]: You don't need to pass in the spec, the use can call `$c->openapi->spec` from inside the helper instead.

File: t/not_implemented_hook.t [hunk 3]
```diff
@@ -0,0 +1,48 @@
+use Mojo::Base -strict;
+use Test::Mojo;
+use Test::More;
+
+use Mojolicious::Lite;
+
+plugin OpenAPI => {
+  url => 'data://main/emulation.json',
+  not_implemented => sub {
+    my ( $c,$spec ) = @_;
```
Comment (jhthorsen) [inline t/not_implemented_hook.t]: Please tidy all the files. There's a default .perltidyrc you can use in the root of this project.

File: lib/Mojolicious/Plugin/OpenAPI.pm [hunk 4]
```diff
@@ -373,6 +375,24 @@ Returns the L<Mojolicious::Controller> object if the input is valid or
 automatically render an error document if not and return false. See
 L</SYNOPSIS> for example usage.
 
+=head2 openapi.not_implemented
```
Comment (jhthorsen) [inline lib/Mojolicious/Plugin/OpenAPI.pm]: Please mark this helper as EXPERIMENTAL:

> This helper is EXPERIMENTAL and subject for change.

**Comment by leejo** [comment]
@jhthorsen - changes made (force pushed to prevent redundant commits)

File: lib/Mojolicious/Plugin/OpenAPI.pm [hunk 5]
```diff
@@ -373,6 +375,18 @@ Returns the L<Mojolicious::Controller> object if the input is valid or
...
+
+  my $response_data = $c->openapi->not_implemented( $openapi_spec );
```
Comment (jhthorsen) [inline lib/Mojolicious/Plugin/OpenAPI.pm]: This helper does not take any input.

**Pull request merged**